### PR TITLE
rework considerations on the choice of hash algorithm 

### PR DESCRIPTION
**Pull request by bc-pi**
for #366 and #359 based largely on text suggested on list by Neil Madden

### CODE REVIEW

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 1]
```diff
@@ -893,19 +893,32 @@ a different salt.
...
-For the security of this scheme, the hash algorithm is required to be preimage resistant and second-preimage
-resistant, i.e., it is infeasible to calculate the salt and claim value that result in
-a particular digest, and, for any salt and claim value pair, it is infeasible to find a different salt and claim value pair that
-result in the same digest, respectively.
-
-Hash algorithms that do not meet the aforementioned requirements MUST NOT be used.
-Inclusion in the "Named Information Hash Algorithm" registry [@IANA.Hash.Algorithms]
+To ensure privacy of claims that are not being selectively disclosed in a given presentation,
+the hash function MUST ensure that it is infeasible to calculate the salt and claim name-value pair
+(or any portion thereof) that results in a particular digest. This implies the hash function MUST
```
Comment (Sakurann) [inline draft-ietf-oauth-selective-disclosure-jwt.md:898]: `claim name-value pair` is somewhat confusing and is not used elsewhere in the spec for this purpose.

```suggestion
the hash function MUST ensure that it is infeasible to calculate one or more of the three elements
(salt, claim name, claim value) that results in a particular digest. This implies the hash function MUST
```

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 2]
```diff
@@ -893,19 +893,32 @@ a different salt.
...
+be preimage resistant, but should also not allow an observer to infer any partial information about
+the undisclosed content. In the terminology of cryptographic commitment schemes, the hash function
+MUST be computationally hiding.
+
+To ensure the integrity of selectively disclosable claims, the hash function MUST be second-preimage
+resistant. That is, for any salt and claim name-value pair, it is infeasible to find a different salt
+and claim name-value pair that result in the same digest.
```
Comment (Sakurann) [inline draft-ietf-oauth-selective-disclosure-jwt.md:905]:
```suggestion
resistant. That is, for any combination of salt, claim name and claim value, it is infeasible to find a different combination of salt, claim name and claim value that result in the same digest.
```

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 3]
```diff
@@ -893,19 +893,32 @@ a different salt.
...
 ## Choice of a Hash Algorithm
 
-For the security of this scheme, the hash algorithm is required to be preimage resistant and second-preimage
-resistant, i.e., it is infeasible to calculate the salt and claim value that result in
-a particular digest, and, for any salt and claim value pair, it is infeasible to find a different salt and claim value pair that
-result in the same digest, respectively.
-
-Hash algorithms that do not meet the aforementioned requirements MUST NOT be used.
-Inclusion in the "Named Information Hash Algorithm" registry [@IANA.Hash.Algorithms]
+To ensure privacy of claims that are not being selectively disclosed in a given presentation,
```
Comment (Sakurann) [inline draft-ietf-oauth-selective-disclosure-jwt.md:896]:
```suggestion
To ensure privacy of claims that are selectively disclosable, but are not being disclosed in a given presentation,
```

File: draft-ietf-oauth-selective-disclosure-jwt.md [hunk 4]
```diff
@@ -893,19 +893,32 @@ a different salt.
...
+MUST be computationally hiding.
+
+To ensure the integrity of selectively disclosable claims, the hash function MUST be second-preimage
+resistant. That is, for any salt and claim name-value pair, it is infeasible to find a different salt
+and claim name-value pair that result in the same digest.
+
+The hash function SHOULD also be collision resistant. Although not essential to the anticipated uses of
+SD-JWT, without collision resistance an Issuer may be able to find multiple disclosures that have the
+same hash value. The signature over the SD-JWT would not then commit the Issuer to the contents of the
+JWT, which is surprising. Where this is a concern, the collision resistance of the hash function SHOULD
```
Comment (Sakurann) [inline draft-ietf-oauth-selective-disclosure-jwt.md:910]: `, which is surprising` felt out of place, at least to me.

```suggestion
same hash value. In which case, the signature over the SD-JWT would not then commit the Issuer to the contents of the
JWT. Where this is a concern, the collision resistance of the hash function SHOULD
```

**Review by Sakurann — CHANGES_REQUESTED** [review]
made suggestions

**Review by danielfett — COMMENTED** [review]
Thank you, LGTM with Kristina's proposed changes.

File: draft-ietf-oauth-selective-disclosure-jwt.md (hunk 1 above)
Comment (danielfett) [reply draft-ietf-oauth-selective-disclosure-jwt.md:898]:
```suggestion
the hash function MUST ensure that it is infeasible to calculate any portion of the three elements
(salt, claim name, claim value) from a particular digest. This implies the hash function MUST
```

**Review by danielfett — APPROVED** [review]

**Review by Sakurann — APPROVED** [review]

**Pull request merged**